test/python/transpiler/test_echo_rzx_weyl_decomposition.py

```diff
@@ -0,0 +1,203 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test the EchoRZXWeylDecomposition pass"""
+import unittest
+from math import pi
+import numpy as np
+
+from qiskit import QuantumRegister, QuantumCircuit
+
+from qiskit.transpiler.passes.optimization.echo_rzx_weyl_decomposition import (
+    EchoRZXWeylDecomposition,
+)
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.test import QiskitTestCase
+from qiskit.test.mock import FakeParis
+
+import qiskit.quantum_info as qi
+
+from qiskit.quantum_info.synthesis.two_qubit_decompose import (
+    TwoQubitWeylDecomposition,
+)
+
+
+class TestEchoRZXWeylDecomposition(QiskitTestCase):
+    """Tests the EchoRZXWeylDecomposition pass."""
+
+    def setUp(self):
+        super().setUp()
+        self.backend = FakeParis()
+
+    def test_rzx_number_native_weyl_decomposition(self):
+        """Check the number of RZX gates for a hardware-native cx"""
+        qr = QuantumRegister(2, "qr")
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[0], qr[1])
+
+        unitary_circuit = qi.Operator(circuit).data
+
+        after = EchoRZXWeylDecomposition(self.backend)(circuit)
+
+        unitary_after = qi.Operator(after).data
+
+        self.assertTrue(np.allclose(unitary_circuit, unitary_after))
+
+        # check whether after circuit has correct number of rzx gates
+        self.assertRZXgates(unitary_circuit, after)
+
+    def test_non_native_weyl_decomposition(self):
+        """Check the number of RZX gates for a non-hardware-native rzz"""
+        theta = pi / 9
+        qr = QuantumRegister(2, "qr")
+        circuit = QuantumCircuit(qr)
+        circuit.rzz(theta, qr[1], qr[0])
+
+        unitary_circuit = qi.Operator(circuit).data
+
+        dag = circuit_to_dag(circuit)
+        pass_ = EchoRZXWeylDecomposition(self.backend)
+        after = dag_to_circuit(pass_.run(dag))
+
+        unitary_after = qi.Operator(after).data
+
+        self.assertTrue(np.allclose(unitary_circuit, unitary_after))
+
+        # check whether after circuit has correct number of rzx gates
```

Comment:
I don't feel this assertion is really testing what you intend to test: turning each CX in a 3 CX decomposition into a pair of RZXes would also pass.

Reply:
Perhaps I misunderstand the comment but I would expect that turning each CX into three CX gates should also pass. I.e.
```
                                            
q_0: ──■────■────■─────────────■────■────■──
     ┌─┴─┐┌─┴─┐┌─┴─┐┌───────┐┌─┴─┐┌─┴─┐┌─┴─┐
q_1: ┤ X ├┤ X ├┤ X ├┤ Rz(1) ├┤ X ├┤ X ├┤ X ├
     └───┘└───┘└───┘└───────┘└───┘└───┘└───┘
```
is equivalent to
```
q_0: ──■─────────────■──
     ┌─┴─┐┌───────┐┌─┴─┐
q_1: ┤ X ├┤ Rz(1) ├┤ X ├
     └───┘└───────┘└───┘
```
I expect that both circuits would result in a RZX echoed decomposition with two RZX gates. Am I missing something?

Reply:
This assertion is checking that the output satisfies a niceness property: it contains no more 2Q gates than expected. The PR is intended to introduce a less wasteful decomposition, where "less wasteful" defined _differently_ from 2Q gate count — in fact, the old decomposition, trivially rewritten CX-by-CX into echoed RZXes, has the same 2Q gate count in each of the cases covered by the test. Because of that overlap, I would prefer that the test more directly check that the specific relevant kind of niceness is satisfied.

Reply:
Thanks for clarifying that makes sense.